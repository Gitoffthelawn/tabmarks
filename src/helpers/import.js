import bookmarksHelper from './bookmarks';

export default {

  importTabGroupsJson(json, errorCallback) {
    let data;
    try {
      data = JSON.parse(json);
      if (Array.isArray(data.windows)) {
        return data.windows.reduce((result, windowData) =>
          result.then(() => this.importWindows(windowData, errorCallback)),
        Promise.resolve());
      }
      return Promise.reject('Invalid JSON structure');
    } catch (e) {
      return Promise.reject('Unable to parse JSON', e);
    }
  },

  importWindows(data, errorCallback) {
    const groups = this.getGroupsData(data);
    const tabs = this.getTabsData(data);

    if (groups) {
      groups.forEach((group, i) => {
        groups[i].tabs = tabs.filter(t => t.groupId === group.id);
      });
      return this.importGroups(groups, errorCallback);
    }

    return Promise.resolve();
  },

  importGroups(groups, errorCallback) {
    return groups.reduce((result, group) =>
      result.then(() => bookmarksHelper.createFolder(group.title)
        .then(folder => this.importTabs(group.tabs, folder, errorCallback)))
        .catch((error) => {
          if (errorCallback) {
            errorCallback({ type: 'folder', name: group.title, error });
          }
        }),
      Promise.resolve());
  },

  importTabs(tabs, folder, errorCallback) {
    return tabs.reduce((result, tab) =>
      result.then(() => bookmarksHelper.create(tab.title, tab.url, folder.id))
        .catch((error) => {
          if (errorCallback) {
            errorCallback({ type: 'bookmark', name: tab.url, error });
          }
        }),
    Promise.resolve());
  },

  getGroupsData(data) {
    if (data && data.extData && data.extData['tabview-group']) {
      try {
        const groups = JSON.parse(data.extData['tabview-group']);
        return Object.keys(groups).map(id => ({ id: parseInt(id, 10), title: groups[id].title }));
      } catch (e) {
        return null;
      }
    }
    return null;
  },

  getTabsData(data) {
    if (Array.isArray(data.tabs)) {
      return data.tabs
        .map(this.getTabData)
        .filter(t => t && t.url && t.url.indexOf('about:') !== 0);
    }
    return [];
  },

  getTabData(data) {
    if (Array.isArray(data.entries) && data.entries.length > 0) {
      const entry = data.entries[0];
      if (data && data.extData && data.extData['tabview-tab']) {
        try {
          const groupData = JSON.parse(data.extData['tabview-tab']);
          if (groupData && groupData.groupID != null) {
            return {
              title: entry.title,
              url: entry.url,
              groupId: parseInt(groupData.groupID, 10),
            };
          }
        } catch (e) {
          return null;
        }
      }
    }
    return null;
  },

};
