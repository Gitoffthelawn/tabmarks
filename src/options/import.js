if (!window.tm) window.tm = {};

tm.import = {

  importTabGroupsJson(json) {
    let data;
    try {
      data = JSON.parse(json);
      if (Array.isArray(data.windows)) {
        return data.windows.reduce((result, windowData) =>
          result.then(() => this.importWindows(windowData)),
        Promise.resolve());
      }
      return Promise.reject('Invalid JSON structure');
    } catch (e) {
      return Promise.reject('Unable to parse JSON', e);
    }
  },

  importWindows(data) {
    const groups = this.getGroupsData(data);
    const tabs = this.getTabsData(data);

    if (groups) {
      groups.forEach((group, i) => {
        groups[i].tabs = tabs.filter(t => t.groupId === group.id);
      });
      return this.importGroups(groups);
    }

    return Promise.resolve();
  },

  importGroups(groups) {
    return groups.reduce((result, group) =>
      result.then(() => tm.bookmarks.createFolder(group.title)
        .then(folder => this.importTabs(group.tabs, folder)),
      error => console.error(`Could not create folder for group "${group.title}"`, error)),
    Promise.resolve());
  },

  importTabs(tabs, folder) {
    return tabs.reduce((result, tab) =>
      result.then(() => tm.bookmarks.create(tab.title, tab.url, folder.id))
        .catch((error) => {
          console.error(`Could not create bookmark "${tab.url}" in group "${folder.id}"`, error);
        }),
    Promise.resolve());
  },

  getGroupsData(data) {
    if (data && data.extData && data.extData['tabview-group']) {
      try {
        const groups = JSON.parse(data.extData['tabview-group']);
        return Object.keys(groups).map(id => ({ id: parseInt(id, 10), title: groups[id].title }));
      } catch (e) {
        return null;
      }
    }
    return null;
  },

  getTabsData(data) {
    if (Array.isArray(data.tabs)) {
      return data.tabs
        .map(this.getTabData)
        .filter(t => t && t.url && t.url.indexOf('about:') !== 0);
    }
    return [];
  },

  getTabData(data) {
    if (Array.isArray(data.entries) && data.entries.length > 0) {
      const entry = data.entries[0];
      if (data && data.extData && data.extData['tabview-tab']) {
        try {
          const groupData = JSON.parse(data.extData['tabview-tab']);
          if (groupData && groupData.groupID != null) {
            return {
              title: entry.title,
              url: entry.url,
              groupId: parseInt(groupData.groupID, 10),
            };
          }
        } catch (e) {
          return null;
        }
      }
    }
    return null;
  },

};
