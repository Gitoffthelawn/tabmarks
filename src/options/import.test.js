require('./import.js');

const json = '{"version":["tabGroups",1],"session":{"lastUpdate":1503346557406,"startTime":1503346353157,"recentCrashes":0},"windows":[{"tabs":[{"entries":[{"url":"about:addons","title":"Add-ons Manager","charset":"","ID":2,"persist":true}],"lastAccessed":1503346438393,"hidden":false,"attributes":{},"extData":{"tabview-tab":"{\\"groupID\\":1}"},"index":1,"image":"chrome://mozapps/skin/extensions/extensionGeneric-16.png"},{"entries":[{"url":"https://developer.mozilla.org/en-US/","title":"MDN Web Docs","charset":"UTF-8","ID":1,"persist":true}],"lastAccessed":1503346449124,"hidden":false,"attributes":{},"extData":{"tabview-tab":"{\\"groupID\\":1}"},"index":1,"image":"https://developer.cdn.mozilla.net/static/img/favicon32.e1ca6d9bb933.png"},{"entries":[{"url":"https://www.mozilla.org/en-US/firefox/?utm_medium=referral&utm_source=getfirefox-com","title":"A better, faster, private browser for today | Firefox","charset":"UTF-8","ID":1,"persist":true}],"lastAccessed":1503346523593,"hidden":false,"attributes":{},"extData":{"tabview-tab":"{\\"groupID\\":1}"},"index":1,"image":"https://www.mozilla.org/media/img/firefox/favicon.dc6635050bf5.ico"},{"entries":[{"url":"https://github.com/","title":"The world\'s leading software development platform Â· GitHub","charset":"UTF-8","ID":1,"persist":true}],"lastAccessed":1503346521778,"hidden":true,"attributes":{},"extData":{"tabview-tab":"{\\"groupID\\":3,\\"active\\":true}"},"index":1,"image":"https://assets-cdn.github.com/favicon.ico"},{"entries":[{"url":"about:tabgroups#session","title":"Tab Groups Options","charset":"","ID":9,"persist":true}],"lastAccessed":1503346557405,"hidden":false,"attributes":{},"extData":{"tabview-tab":"{\\"groupID\\":1}"},"index":1}],"extData":{"tabview-group":"{\\"1\\":{\\"bounds\\":{\\"left\\":15,\\"top\\":5,\\"width\\":1237.285,\\"height\\":658.996},\\"slot\\":1,\\"userSize\\":null,\\"stackTabs\\":true,\\"showThumbs\\":true,\\"showUrls\\":true,\\"tileIcons\\":true,\\"catchOnce\\":true,\\"catchRules\\":\\"\\",\\"title\\":\\"Group A\\",\\"id\\":1},\\"3\\":{\\"bounds\\":{\\"left\\":20,\\"top\\":20,\\"width\\":250,\\"height\\":200},\\"slot\\":2,\\"userSize\\":null,\\"stackTabs\\":true,\\"showThumbs\\":true,\\"showUrls\\":true,\\"tileIcons\\":true,\\"catchOnce\\":true,\\"catchRules\\":\\"\\",\\"title\\":\\"Group B\\",\\"id\\":3}}","tabview-groups":"{\\"nextID\\":4,\\"activeGroupId\\":1,\\"activeGroupName\\":\\"Group A\\",\\"totalNumber\\":2}"}}]}';

describe('tm.import', () => {
  describe('#importTabGroupsJson', () => {
    beforeEach(() => {
      tm.bookmarks = {
        createFolder: jest.fn().mockImplementation((title) => {
          if (title === 'Group A') {
            return Promise.resolve({ id: 1 });
          } else if (title === 'Group B') {
            return Promise.resolve({ id: 2 });
          }
          return Promise.reject('Unexpected folder title');
        }),
        create: jest.fn().mockImplementation(() => Promise.resolve()),
      };
    });

    test('rejects if invalid JSON data', (done) => {
      tm.import.importTabGroupsJson('foo').catch(done);
    });

    test('create bookmarks from JSON data', () =>
      tm.import.importTabGroupsJson(json).then(() => {
        expect(tm.bookmarks.createFolder.mock.calls).toMatchSnapshot();
        expect(tm.bookmarks.create.mock.calls).toMatchSnapshot();
      }));
  });
});
