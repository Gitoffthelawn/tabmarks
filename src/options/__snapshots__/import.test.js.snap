// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tm.import #importTabGroupsJson create bookmarks from JSON data 1`] = `
Array [
  Array [
    "Group A",
  ],
  Array [
    "Group B",
  ],
]
`;

exports[`tm.import #importTabGroupsJson create bookmarks from JSON data 2`] = `
Array [
  Array [
    "MDN Web Docs",
    "https://developer.mozilla.org/en-US/",
    1,
  ],
  Array [
    "A better, faster, private browser for today | Firefox",
    "https://www.mozilla.org/en-US/firefox/?utm_medium=referral&utm_source=getfirefox-com",
    1,
  ],
  Array [
    "The world's leading software development platform Â· GitHub",
    "https://github.com/",
    2,
  ],
]
`;
